README.md
# mustache.el [![Coverage Status](https://coveralls.io/repos/github/Wilfred/mustache.el/badge.svg)](https://coveralls.io/github/Wilfred/mustache.el)
#### *a mustache templating library in Emacs Lisp*

Targeting [v.1.0.2](https://github.com/mustache/spec/tree/v1.0.2) of Mustache.

## Example usage

``` emacs-lisp
(require 'mustache)
(require 'ht) ;; hash table library

(let ((context (ht ("name" "J. Random user"))))
  ;; evaluates to: "Hello J. Random user!"
  (mustache-render "Hello {{name}}!" context))
```

### (Optional) Without ht

You're not forced to use `ht`, it's just an easier way of creating
hash tables. You can use Emacs' reader syntax for hash tables instead:
      
``` emacs-lisp
(require 'mustache)

(let ((context
       #s(hash-table test equal data ("name" "J. Random user"))))
  ;; evaluates to: "Hello J. Random user!"
  (mustache-render "Hello {{name}}!" context))
```

Note that hash tables default to using `eql` as the key comparison
function. You must set it to `equal` since mustache.el uses hash
tables with string keys.

### Keywords in context

You can use keywords in contexts, which allows you to skip setting the
key comparison function.

``` emacs-lisp
(require 'mustache)

(let ((mustache-key-type 'keyword)
      (context
       #s(hash-table data (:name "J. Random user"))))
  ;; evaluates to: "Hello J. Random user!"
  (mustache-render "Hello {{name}}!" context))
```

## Implemented mustache features

Basic variable interpolation:

``` emacs-lisp
(mustache-render
 "Coded with {{language}}!"
 (ht ("language" "elisp"))) ;; "Coded with elisp!"
```
     
Blocks with booleans:

``` emacs-lisp
(mustache-render
 "{{#is-sunny}}Looks nice today.{{/is-sunny}}"
 (ht ("is-sunny" t))) ;; "Looks nice today."

(mustache-render
 "{{#is-sunny}}Looks nice today.{{/is-sunny}}"
 (ht ("is-sunny" nil))) ;; ""
```
     
Blocks with hash tables:

``` emacs-lisp
(mustache-render
 "{{#user}}{{name}}{{/user}}"
 (ht ("user"
      (ht ("name" "Wilfred"))))) ;; "Wilfred"
```
     
Blocks with lists:

``` emacs-lisp
(mustache-render
 "{{#some-list}}{{item}}{{/some-list}}"
 (ht ("some-list"
      (list
       (ht ("item" "a"))
       (ht ("item" "b"))
       (ht ("item" "c")))))) ;; "abc"
```

Inverted blocks:

``` emacs-lisp
(mustache-render
 "{{^is-sunny}}Take an umbrella!{{/is-sunny}}"
 (ht ("is-sunny" nil))) ;; "Take an umbrella!"

(mustache-render
 "{{^is-sunny}}Take an umbrella!{{/is-sunny}}"
 (ht ("is-sunny" t))) ;; ""
```

Mustache variables are escaped:

``` emacs-lisp
(mustache-render
 "{{info}}"
 (ht ("info" "<p>We use mustache</p>"))) ;; "&lt;p&gt;We use mustache&lt;/p&gt;"
```

Unless explicitly marked as safe:

``` emacs-lisp
(mustache-render
 "{{{info}}}"
 (ht ("info" "<p>We use mustache</p>"))) ;; "<p>We use mustache</p>"

(mustache-render
 "{{& info }}"
 (ht ("info" "<p>We use mustache</p>"))) ;; "<p>We use mustache</p>"
```

Comments:

``` emacs-lisp
(mustache-render
 "hello{{! world }}"
 (ht)) ;; "hello"
```

Partials:

``` emacs-lisp
;; assuming ~/projects/mustache.el/test.mustache exists
;; and contains "hello {{user}}"
(let ((mustache-partial-paths (list "~/projects/mustache.el")))
  (mustache-render
   "{{> test}}"
   (ht ("user" "wilfred")))) ;; "hello wilfred"
```

Changing delimeters:

``` emacs-lisp
(mustache-render
 "{{=<% %>=}}<% style %>"
 (ht ("style" "ERB style!"))) ;; "ERB style!"
```

Lambdas:

``` emacs-lisp
(mustache-render
 "{{#wrapped}}{{language}} is great.{{/wrapped}}"
 (ht ("language" "elisp")
     ("wrapped"
      (lambda (template context)
        (concat "<b>" (mustache-render template context) "</b>")))))
;; "<b>elisp is great.</b>"
```

Error checking on invalid sections:

``` emacs-lisp
(mustache-render
 "{{#outer}}{{#inner}}mismatched!{{/outer}}{{/inner}}"
 (ht)) ;; error "Mismatched brackets: You closed a section with inner, but it wasn't open"
```

## Todo:

* Errors on unclosed tags
* Optional error on missing variables from the context
* Whitespace (in)sensitivity for windows newlines
* Run full specification test suite

### Running tests

Within Emacs:

    M-x mustache-run-tests

Or from a command line (you need Cask installed):

    $ cask
    $ cask exec ert-runner

### Roadmap

v1.0 -- Pass the full mustache v1.0.2 specification tests (excluding
optional parts).

## Other templating projects

* The `format` function (quick and dirty!)
* [esxml](https://github.com/tali713/esxml)
* [elnode](https://github.com/nicferrier/elnode) (docs [here](https://github.com/nicferrier/elnode#sending-files))
* `s-format` from [s.el](https://github.com/magnars/s.el)
* [xmlgen](https://github.com/philjackson/xmlgen)


CHANGELOG.md
* v0.24 Fixed an issue with `{{{foo}}}` after changing delimiters.
* v0.23 Added the variable `mustache-key-type` which allows contexts
  with keyword for keys.
* v0.22 Fixed rendering of contexts with numeric values and
  `{{&escaped blocks}}`
* v0.21 Fixed rendering of contexts where the hash table contained
  number values.
* v0.20 improved whitespace insensitivity for `{{^blocks}}` and
  `{{!comments}}`
* v0.19 Fixed a bug where comment tags containing = were treated as
  changing the delimeters.
* v0.18 Improved whitespace (in)sensitivity
* v0.17 Fixed a bug where partial templates weren't found when
  `default-directory` was set.
* v0.16 Implemented lambdas.
* v0.15 Fixed partials not being recursively rendered, and added error
  checking for `mustache-partial-path` not being a list
* v0.14 Implemented changing tag delimeters e.g. `{{=<% %>=}}`
* v0.13 Improved whitespace sensitivity for `{{ variable }}`
  interpolation
* v0.12 Implemented partials `{{> foo}}`
* v0.11 Fixed nested contexts allowing access to the parent context
* v0.10 Whitespace sensitive now matches the mustache spec in more cases
* v0.9 Fixed rendering of `{{#blocks}}` when the context value is a
  hash table.
* v0.8 Removed dependency on `with-namespace` in favour of internal
  functions/variables being of the form `mst--foo`.
* v0.7 templates with malformed sections now throw an error. For
  example: `{{#foo}}{{/bar}}` and `{{#foo}}`.
* v0.6 `{{#blocks}}` now optionally support using vectors instead of
  lists
* v0.5 added rendering of `{{{blocks}}}`
* v0.4 internal functions are now named mustache--foo instead of
  mustache/foo due to use of the `with-namespace` macro
* v0.3 internal refactoring
* v0.2 added rendering `{{#blocks}}` when the context contains a list
  of hash tables
* v0.1 basic implementation of a subset of mustache


mustache-spec.el
(require 'json)
(require 'ht)
(require 'dash)
(require 'mustache)
(require 'ert)
(require 's)
(eval-when-compile (require 'cl))

(eval-and-compile
  (defun mst--read-json-path (path)
    "Open the JSON file at PATH, parse it, and return a hash table of its contents."
    (let ((json-object-type 'hash-table)
          (json-array-type 'list)
          (json-false nil))
      (with-temp-buffer
        (insert-file-contents-literally path)
        (json-read))))

  (defun mst--ert-from-spec (spec-case)
    "Return a quoted ert test from a hash-table SPEC-CASE
describing expected behaviour"
    (let* ((spec-name (ht-get spec-case "name"))
           (test-name (intern (format "mustache-spec-%s" (s-replace " " "-" spec-name))))
           (description (ht-get spec-case "desc"))
           (template (ht-get spec-case "template"))
           (context (ht-get spec-case "data"))
           (expected (ht-get spec-case "expected")))
      `(ert-deftest ,test-name ()
         ,description
         (should
          (equal
           ,expected (mustache-render ,template ,context)))))))

(defmacro ert-tests-from-path (path)
  "Open the JSON file at PATH, and return a list of ert test cases."
  `(progn
     ,@(--map (mst--ert-from-spec it)
              (ht-get (mst--read-json-path path) "tests"))))

(ert-tests-from-path "spec/comments.json")
(ert-tests-from-path "spec/delimeters.json")
(ert-tests-from-path "spec/interpolation.json")
(ert-tests-from-path "spec/inverted.json")
;; todo: partials
(ert-tests-from-path "spec/sections.json")

(defun mustache-test-spec ()
  (interactive)
  (ert-run-tests-interactively "mustache-spec-"))


mustache-tests.el
;;; mustache-test.el --- Mustache test module

;; Copyright (C) 2013 Wilfred Hughes

;; Author: Wilfred Hughes <me@wilfred.me.uk>
;; URL: https://github.com/Wilfred/mustache.el

;; This file is part of mustache.el.

;; mustache.el is free software: you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; mustache.el is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with mustache.el.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Mustache test module.

;;; Code:

(require 'mustache)
(require 'ert)
(require 'ht)

(ert-deftest mustache-test-simple-string ()
  (should (equal "foo" (mustache-render "foo" nil))))

(ert-deftest mustache-test-variable ()
  (let ((context (ht ("blah" "bar"))))
    (should
     (equal
      "foo bar"
      (mustache-render "foo {{blah}}" context)))))

(ert-deftest mustache-malformed-tag ()
  "Don't crash if we have a malformed tag.
Arguably we could error, but mustache generally errs on error
tolerance, e.g. allowing missing variables."
  (mustache-render "}} foo {{" (ht)))

(ert-deftest mustache-test-variable-number ()
  (let ((context (ht ("blah" 2))))
    (should
     (equal
      "foo 2"
      (mustache-render "foo {{blah}}" context)))))

(ert-deftest mustache-test-unescaped-variable-number ()
  (let ((context (ht ("blah" 2))))
    (should
     (equal
      "foo 2"
      (mustache-render "foo {{{blah}}}" context)))))

(ert-deftest mustache-test-variable-whitespace ()
  (let ((context (ht ("blah" "bar"))))
    (should
     (equal
      "foo bar"
      (mustache-render "foo {{ blah }}" context)))))

(ert-deftest mustache-test-variable-missing ()
  (let ((context (ht-create)))
    (should
     (equal
      "foo "
      (mustache-render "foo {{blah}}" context)))))

(ert-deftest mustache-test-keyword-variable ()
  (let ((mustache-key-type 'keyword)
        (context (ht (:blah "bar"))))
    (should
     (equal
      "foo bar"
      (mustache-render "foo {{blah}}" context)))))

(ert-deftest mustache-test-section-inner-whitespace ()
  (should
   (equal
    "="
    (mustache-render "{{# boolean }}={{/ boolean }}" (ht ("boolean" t))))))

(ert-deftest mustache-test-standalone-lines ()
  (should
   (equal "| This Is
|
| A Line
"
          (mustache-render "| This Is
{{#boolean}}
|
{{/boolean}}
| A Line
" (ht ("boolean" t))))))

(ert-deftest mustache-test-standalone-lines-inverted ()
  (should
   (equal "| This Is
|
| A Line
"
          (mustache-render "| This Is
{{^boolean}}
|
{{/boolean}}
| A Line
" (ht ("boolean" nil))))))

(ert-deftest mustache-test-standalone-lines-leading-whitespace ()
  (should
   (equal "| This Is
|
| A Line
"
          (mustache-render "| This Is
  {{#boolean}}
|
  {{/boolean}}
| A Line
" (ht ("boolean" t))))))

(ert-deftest mustache-test-standalone-lines-comments ()
  (should
   (equal "| This Is
|
| A Line
"
          (mustache-render "| This Is
|
  {{!comment}}
| A Line
" (ht)))))

(ert-deftest mustache-test-extra-section-close ()
  (should-error
   (mustache-render "{{/blah}}" (ht-create))))

(ert-deftest mustache-test-extra-section-open ()
  (should-error
   (mustache-render "{{#blah}}" (ht-create))))

(ert-deftest mustache-test-mismatched-section-close ()
  (should-error
   (mustache-render "{{#blah}}{{/foo}}" (ht-create))))

(ert-deftest mustache-test-variable-escaped ()
  (let ((context (ht ("blah" "<bar> &baz ' \""))))
    (should
     (equal
      "&lt;bar&gt; &amp;baz &#39; &quot;"
      (mustache-render "{{blah}}" context)))))

(ert-deftest mustache-test-section-hash ()
  (let ((context (ht ("user"  (ht ("name" "bob"))))))
    (should
     (equal
      "bob"
      (mustache-render "{{#user}}{{name}}{{/user}}" context)))))

(ert-deftest mustache-test-section-hash-nested ()
  (let ((context (ht
                  ("foo" "bar")
                  ("user"  (ht ("name" "bob"))))))
    (should
     (equal
      "barbob"
      (mustache-render "{{#user}}{{foo}}{{name}}{{/user}}" context)))))

(ert-deftest mustache-test-section-list ()
  (let ((context (ht ("users" (list (ht ("name" "bob"))
                                    (ht ("name" "chris")))))))
    (should
     (equal
      "bobchris"
      (mustache-render "{{#users}}{{name}}{{/users}}" context)))))

(ert-deftest mustache-test-section-with-vector ()
  "Vectors should behave the same as a list in a context."
  (let ((context (ht ("users" `[,(ht ("name" "bob"))
                                ,(ht ("name" "chris"))]))))
    (should
     (equal
      "bobchris"
      (mustache-render "{{#users}}{{name}}{{/users}}" context)))))

(ert-deftest mustache-test-unescaped ()
  (let ((context (ht ("blah" "<bar>"))))
    (should
     (equal
      "<bar>"
      (mustache-render "{{& blah}}" context)))))

(ert-deftest mustache-test-triple-mustache ()
  (let ((context (ht ("blah" "<bar>"))))
    (should
     (equal
      "<bar>"
      (mustache-render "{{{blah}}}" context)))))

(ert-deftest mustache-test-triple-mustache-custom-delimiter ()
  (let ((context (ht ("name" "wilfred"))))
    (should
     (equal
      "{{{name}}}"
      (mustache-render "{{=<< >>=}}{{{name}}}" context)))))

(ert-deftest mustache-test-conditional-true ()
  (let ((context (ht ("yes" 't))))
    (should
     (equal
      "foo bar"
      (mustache-render "foo {{#yes}}bar{{/yes}}" context)))))

(ert-deftest mustache-test-conditional-false ()
  (let ((context (ht ("no" nil))))
    (should
     (equal
      "foo "
      (mustache-render "foo {{#no}}bar{{/no}}" context)))))

(ert-deftest mustache-test-conditional-keyword ()
  (let ((mustache-key-type 'keyword)
        (context (ht (:no nil))))
    (should
     (equal
      "foo "
      (mustache-render "foo {{#no}}bar{{/no}}" context)))))

(ert-deftest mustache-test-inverted ()
  (let ((context (ht ("no" nil))))
    (should
     (equal
      "foo bar"
      (mustache-render "foo {{^no}}bar{{/no}}" context)))))

(ert-deftest mustache-test-comment ()
  (let ((context (ht-create)))
    (should
     (equal
      ""
      (mustache-render "{{! whatever}}" context)))))

(ert-deftest mustache-test-after-comment ()
  "Ensure we render tags after comments.
Regression test for https://github.com/Wilfred/mustache.el/issues/4."
  (let ((context  (ht ("b" "hello"))))
    (should
     (equal
      "hello"
      (mustache-render
       "{{!<script src=\"somescript.js\" async=\"async\"}}{{b}}" context)))))

(ert-deftest mustache-test-partial ()
  (let ((mustache-partial-paths (list default-directory)))
    (should
     (equal
      "hello world"
      (mustache-render "{{> partial }}" (ht))))))

(ert-deftest mustache-test-partial-path ()
  "Ensure we get the correct partial path, regardless of `default-directory'.
Regression test for https://github.com/Wilfred/mustache.el/pull/3"
  (let ((mustache-partial-paths (list default-directory))
        (default-directory "/"))
    (should
     (equal
      "hello world"
      (mustache-render "{{> partial }}" (ht))))))

(ert-deftest mustache-test-partial-rendered ()
  "Test that we render the contents of the partial as a mustache template."
  (let ((mustache-partial-paths (list default-directory)))
    (should
     (equal
      "hello world"
      (mustache-render
       "{{> partial-with-variables }}"
       (ht ("thing" "world")))))))

(ert-deftest mustache-test-change-delimeter ()
  (should
   (equal
    "bar"
    (mustache-render "{{=<< >>=}}<< foo >>" (ht ("foo" "bar"))))))

(ert-deftest mustache-test-lambda ()
  (should
   (equal
    "<b>Willy is awesome.</b>"
    (mustache-render
     "{{#wrapped}}{{name}} is awesome.{{/wrapped}}"
     (ht ("name" "Willy")
         ("wrapped"
          (lambda (template context)
            (concat "<b>" (mustache-render template context) "</b>"))))))))

(defun mustache-run-tests ()
  (interactive)
  (ert-run-tests-interactively "mustache-test-"))

(provide 'mustache-tests)
;;; mustache-tests.el ends here


test/test-helper.el
;;; test-helper.el --- Helper for tests              -*- lexical-binding: t; -*-

;; Copyright (C) 2020 Wilfred Hughes

;; Author:  <me@wilfred.me.uk>

;;; Code:

(require 'ert)
(require 'f)

(let ((mustache-dir (f-parent (f-dirname (f-this-file)))))
  (add-to-list 'load-path mustache-dir))

(require 'undercover)
(undercover "*.el" (:exclude "*-tests.el"))

;;; test-helper.el ends here


mustache-pkg.el
;; This file is part of mustache.el.
;;
;; mustache.el is free software: you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; mustache.el is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with mustache.el.  If not, see <https://www.gnu.org/licenses/>.

(define-package
  "mustache" "0.20"
  "a mustache templating library in emacs lisp"
  '((ht "0.9") (s "1.3.0") (dash "1.2.0")))


mustache.el
;;; mustache.el --- Mustache templating library in emacs lisp

;; Copyright (C) 2013 Wilfred Hughes

;; Author: Wilfred Hughes <me@wilfred.me.uk>
;; Version: 0.24
;; Keywords: convenience mustache template
;; Package-Requires: ((ht "0.9") (s "1.3.0") (dash "1.2.0"))
;; URL: https://github.com/Wilfred/mustache.el

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; See documentation at https://github.com/Wilfred/mustache.el

;; Note on terminology: We treat mustache templates as a sequence of
;; strings (plain text), and tags (anything wrapped in delimeters:
;; {{foo}}). A section is a special tag that requires closing
;; (e.g. {{#foo}}{{/foo}}).

;; We treat mustache templates as if they conform to a rough grammar:

;; TEMPLATE = plaintext | TAG | SECTION | TEMPLATE TEMPLATE
;; SECTION = OPEN-TAG TEMPLATE CLOSE-TAG
;; TAG = "{{" text "}}"

;; Public functions are of the form `mustache-FOO`, private
;; functions/variables are of the form `mst--FOO`.

;;; Code:

(require 's)
(require 'ht)
(require 'dash)
(require 'cl-lib)

;; todo: add flag to set tolerance of missing variables
(defun mustache-render (template context)
  "Render a mustache TEMPLATE with hash table CONTEXT."
  (-> template
      mst--lex
      mst--clean-whitespace
      mst--parse
      (mst--render-section-list context)))

(defvar mustache-partial-paths nil
  "A list of paths to be searched for mustache partial templates (files ending .mustache).")

(defvar mustache-key-type 'string
  "What type of key we expect in contexts.
Can take the value 'string or 'keyword.

For 'string we expect contexts of the form:
#s(hash-table data (\"name\" \"J. Random user\"))

for 'keyword we expect contexts of the form:
#s(hash-table data (:name \"J. Random user\"))")

;;; Lexing 

(defun mst--lex (template)
  "Iterate through TEMPLATE, splitting {{ tags }} and bare strings.
We return a list of lists: ((:text \"foo\") (:tag \"variable-name\"))"
  (let ((open-delimeter "{{")
        (close-delimeter "}}")
        (lexemes nil))
    (while (not (s-equals? template ""))
      (let* ((open-index (s-index-of open-delimeter template))
             (close-index (s-index-of close-delimeter template)))
        ;; todo: error if we have an open and no close
        (if (and open-index close-index
                 (< open-index close-index))
            ;; We have a well-formed tag.
            (progn
              ;; If we have a triple delimiter {{{foo}}}, we want to
              ;; consider the inner content as "{foo}", so we need to
              ;; increment the close index.
              (when
                  (and (< close-index (- (length template) 2))
                       (eq (elt template (+ close-index 2)) ?\}))
                (setq close-index (1+ close-index)))
              
              (let ((between-delimeters
                     (substring template (+ open-index (length open-delimeter)) close-index))
                    (continue-from-index (+ close-index (length close-delimeter))))
                ;; save the string before the tag
                (when (> open-index 0)
                  (push (list :text (substring template 0 open-index)) lexemes))

                ;; if this is a tag that changes delimeters e.g. {{=<< >>=}}
                ;; then set the new open/close delimeter string
                (if (s-matches-p "^=.+ .+=$" between-delimeters)
                    (let* (;; strip leading/trailing =
                           (delimeter-spec (substring between-delimeters 1 -1))
                           (spec-parts (s-split " " delimeter-spec)))
                      (setq open-delimeter (-first-item spec-parts))
                      (setq close-delimeter (-second-item spec-parts)))

                  ;; otherwise it's a normal tag, so save it
                  (push (list :tag (s-trim between-delimeters)) lexemes))

                ;; iterate on the remaining template
                (setq template
                      (substring template continue-from-index))))
          ;; else only plain text left
          (progn
            (push (list :text template) lexemes)
            (setq template "")))))
    (nreverse lexemes)))

(defun mst--clean-whitespace (lexemes)
  "Given a list of LEXEMES, remove whitespace around sections and
comments if they're on their own on a line.  Modifies the original
list."
  ;; iterate over all lexemes:
  (dotimes (i (- (length lexemes) 2))
    ;; find sections that have plain text before and after
    (let ((first (elt lexemes i))
          (second (elt lexemes (+ i 1)))
          (third (elt lexemes (+ i 2))))
      (when (and (mst--text-p first)
                 (or
                  (mst--section-tag-p second)
                  (mst--comment-tag-p second))
                 (mst--text-p third)
                 ;; check the section is on its own line
                 (string-match-p "\n *$" (mst--tag-text first))
                 (string-match-p "^\n" (mst--tag-text third)))
        ;; then we cleanup whitespace
        (setf (elt lexemes i) (mst--no-trailing-newline first)))))
  lexemes)

(defalias 'mst--tag-text '-second-item
  "Returns the text context of a tag.")

(defun mst--no-trailing-newline (lexeme)
  "Replace \"\n\" or \"\n   \" at the end of a plain text LEXEME."
  (list
   :text
   (replace-regexp-in-string "\n *$" "" (mst--tag-text lexeme))))

(defun mst--tag-p (lexeme)
  "Is LEXEME a tag?"
  (eq (car lexeme) :tag))

(defun mst--section-tag-p (lexeme)
  "Is LEXEME a section tag?"
  (or
   (mst--open-section-tag-p lexeme)
   (mst--inverted-section-tag-p lexeme)
   (mst--close-section-tag-p lexeme)))

(defun mst--open-section-tag-p (lexeme)
  "Is LEXEME an open section tag?
See also `mst--inverted-section-tag-p'."
  (and (mst--tag-p lexeme)
       (s-starts-with-p "#" (-second-item lexeme))))

(defun mst--inverted-section-tag-p (lexeme)
  "Is LEXEME an inverted section tag?"
  (and (mst--tag-p lexeme)
       (s-starts-with-p "^" (-second-item lexeme))))

(defun mst--close-section-tag-p (lexeme)
  "Is LEXEME a close section tag?"
  (and (mst--tag-p lexeme)
       (s-starts-with-p "/" (-second-item lexeme))))

(defun mst--comment-tag-p (lexeme)
  "Is LEXEME a comment tag?"
  (and (mst--tag-p lexeme)
       (s-starts-with-p "!" (-second-item lexeme))))

(defun mst--unescaped-tag-p (lexeme)
  "Is LEXEME an unescaped variable tag?"
  (and (mst--tag-p lexeme)
       (let ((text (-second-item lexeme)))
         (or
          (s-starts-with-p "&" text)
          (and
           (s-starts-with-p "{" text)
           (s-ends-with-p "}" text))))))

(defun mst--partial-tag-p (lexeme)
  "Is LEXEME a partial tag?"
  (and (mst--tag-p lexeme)
       (s-starts-with-p ">" (-second-item lexeme))))

(defun mst--section-p (lexeme)
  "Is LEXEME a nested section?"
  (not (atom (car lexeme))))

(defun mst--text-p (lexeme)
  "Is LEXEME plain text?"
  (eq (car lexeme) :text))

;; fixme: assumes the delimeters haven't changed
;; fixme: mst--lex doesn't preserve whitespace
(defun mst--unlex (lexemes)
  "Given a lexed (and optionally parsed) list of LEXEMES,
return the original input string."
  (if lexemes
      (let ((lexeme (car lexemes))
            (rest (cdr lexemes)))
        (cond
         ;; recurse on this section, then the rest
         ((mst--section-p lexeme)
          (concat (mst--unlex lexeme) (mst--unlex rest)))
         ((mst--tag-p lexeme)
          ;; restore the delimeters, then unlex the rest
          (let ((tag-name (-second-item lexeme)))
            (concat "{{" tag-name "}}" (mst--unlex rest))))
         ;; otherwise, it's just raw text
         (t
          (let ((text (-second-item lexeme)))
            (concat text (mst--unlex rest))))))
    ""))

;;; Parsing 

(defvar mst--remaining-lexemes nil
  "Since `mst--parse-inner' recursively calls itself, we need a shared value to mutate.")

(defun mst--parse (lexemes)
  "Given a list LEXEMES, return a list of lexemes nested according to #tags or ^tags."
  (setq mst--remaining-lexemes lexemes)
  (mst--parse-inner))

(defun mst--parse-inner (&optional section-name)
  "Parse `mst--remaining-lexemes', and return a list of lexemes nested according to #tags or ^tags."
  (let (parsed-lexemes
        lexeme)
    (catch 'done
      (while mst--remaining-lexemes
        (setq lexeme (pop mst--remaining-lexemes))
        (cond
         ((mst--open-section-p lexeme)
          ;; recurse on this nested section
          (push (cons lexeme (mst--parse-inner (mst--section-name lexeme)))
                parsed-lexemes))
         ((mst--close-section-p lexeme)
          ;; this is the last tag in this section
          (unless (equal section-name (mst--section-name lexeme))
            (error "Mismatched brackets: You closed a section with %s, but it wasn't open" section-name))
          (push lexeme parsed-lexemes)
          (throw 'done nil))
         (t
          ;; this is just a tag in the current section
          (push lexeme parsed-lexemes)))))
    
    ;; ensure we aren't inside an unclosed section
    (when (and section-name (not (mst--close-section-p lexeme)))
      (error "Unclosed section: You haven't closed %s" section-name))

    (nreverse parsed-lexemes)))

(defun mst--open-section-p (lexeme)
  "Is LEXEME a #tag or ^tag ?"
  (-let [(type value) lexeme]
    (and (eq type :tag)
         (or
          (s-starts-with-p "#" value)
          (s-starts-with-p "^" value)))))

(defun mst--close-section-p (lexeme)
  "Is LEXEME a /tag ?"
  (-let [(type value) lexeme]
    (and (eq type :tag)
         (s-starts-with-p "/" value))))

;;; Rendering 

(defvar mustache-key-type)
(defvar mustache-partial-paths)

;; todo: set flag to set tolerance of missing templates
(defun mst--get-partial (name)
  "Get the first partial whose file name is NAME.mustache, or nil otherwise.
Partials are searched for in `mustache-partial-paths'."
  (cl-block nil
    (unless (listp mustache-partial-paths)
      (error "`mustache-partial-paths' must be a list of paths"))
    (let ((partial-name (format "%s.mustache" name)))
      (dolist (path mustache-partial-paths)
        (-when-let*
            ((partials (directory-files path nil "\\.mustache$"))
             (matching-partial (--first
                                (string-match-p (regexp-quote partial-name) it)
                                partials)))
          (cl-return
           (with-temp-buffer
             (insert-file-contents-literally (expand-file-name matching-partial path))
             (buffer-substring-no-properties (point-min) (point-max)))))))))

(defun mst--render-section-list (sections context)
  "Render a parsed list SECTIONS in CONTEXT."
  (s-join
   ""
   (--map (mst--render-section it context) sections)))

(defun mst--context-get (context variable-name &optional default)
  "Lookup VARIABLE-NAME in CONTEXT, returning DEFAULT if not present."
  (when (eq mustache-key-type 'keyword)
    (setq variable-name (intern (concat ":" variable-name))))
  (ht-get context variable-name default))

(defun mst--tag-name (tag-text)
  "Given a tag {{foo}}, {{& foo}} or {{{foo}}}, return \"foo\"."
  (s-trim
   (cond
    ((s-starts-with-p "&" tag-text)
     (substring tag-text 1))
    ((and (s-starts-with-p "{" tag-text)
          (s-ends-with-p "}" tag-text))
     (substring tag-text 1 -1)))))

(defun mst--render-tag (parsed-tag context)
  "Given PARSED-TAG, render it in hash table CONTEXT."
  (let ((inner-text (mst--tag-text parsed-tag)))
    (cond
     ((mst--comment-tag-p parsed-tag)
      "")
     ((mst--unescaped-tag-p parsed-tag)
      (let ((variable-value (mst--context-get context (mst--tag-name inner-text) "")))
        (when (numberp variable-value)
          (setq variable-value (number-to-string variable-value)))
        variable-value))
     ((mst--partial-tag-p parsed-tag)
      (let ((partial (mst--get-partial (s-trim (substring inner-text 1)))))
        (if partial
            (mustache-render partial context)
          "")))
     (t ;; normal variable
      (let ((variable-value (mst--context-get context inner-text "")))
        (when (numberp variable-value)
          (setq variable-value (number-to-string variable-value)))
        (mst--escape-html variable-value))))))

(defun mst--context-add (table from-table)
  "Return a copy of TABLE where all the key-value pairs in FROM-TABLE have been set."
  (let ((new-table (ht-copy table)))
    (ht-update new-table from-table)
    new-table))

(defun mst--listp (object)
  "Return t if OBJECT is a list.
Unlike `listp', does not return t if OBJECT is a function."
  (and (not (functionp object)) (listp object)))

(defun mst--section-name (section-tag)
  "Get the name of this SECTION-TAG.
E.g. from {{#foo}} to \"foo\"."
  (-> section-tag ;; e.g (:tag "#foo")
      mst--tag-text ;; to "#foo"
      (substring 1) ;; to "foo"
      s-trim))

(defun mst--render-section (parsed-lexeme context)
  "Given PARSED-LEXEME -- a lexed tag, plain text, or a nested list,
render it in CONTEXT."
  (cond ((mst--section-p parsed-lexeme)
         ;; nested section
         (let* ((section-tag (car parsed-lexeme))
                (section-name (mst--section-name section-tag))
                (context-value (mst--context-get context section-name))
                ;; strip section open and close
                (section-contents (-slice parsed-lexeme 1 -1)))
           (cond
            ((mst--open-section-tag-p section-tag)
             (cond
              ;; if the context is a list of hash tables, render repeatedly
              ((or (mst--listp context-value) (vectorp context-value))
               (s-join
                ""
                (--map
                 (mst--render-section-list section-contents (mst--context-add context it))
                 context-value)))
              ;; if the context is a hash table, render in that context
              ((hash-table-p context-value)
               (mst--render-section-list section-contents (mst--context-add context context-value)))
              ;; if the context is a function, call it
              ((functionp context-value)
               (funcall context-value (mst--unlex section-contents) context))
              ;; if it's a truthy value, render in the current context
              (context-value
               (mst--render-section-list section-contents context))
              ;; otherwise, don't render anything
              (t "")))
            ;; render ^tags
            ((mst--inverted-section-tag-p section-tag)
             (if context-value
                 ""
               (mst--render-section-list section-contents context))))))
        ((mst--tag-p parsed-lexeme)
         (mst--render-tag parsed-lexeme context))
        ;; plain text
        (t
         (-second-item parsed-lexeme))))

(defun mst--escape-html (string)
  "Escape HTML in STRING."
  (->> string
       (s-replace "&" "&amp;")
       (s-replace "<" "&lt;")
       (s-replace ">" "&gt;")
       (s-replace "'" "&#39;")
       (s-replace "\"" "&quot;")))

(provide 'mustache)
;;; mustache.el ends here
